stan/math/prim/arr/fun/array_builder.hpp

```diff
@@ -0,0 +1,47 @@
+#ifndef STAN_MATH_PRIM_ARR_FUN_ARRAY_BUILDER_HPP
+#define STAN_MATH_PRIM_ARR_FUN_ARRAY_BUILDER_HPP
+
+#include <stan/math/prim/scal/fun/promote_elements.hpp>
+#include <vector>
+
+namespace stan {
+  namespace math {
+
+    /**
+     * Structure for building up arrays in an expression (rather than
+     * in statements) using an argumentchaining add() method and
+     * a getter method array() to return the result.
+     * Array elements are held in std::vector of type T.
+     *
+     * @tparam T type of array elements
+     */
+    template <typename T>
+    struct array_builder {
+      std::vector<T> x_;
+      array_builder() : x_() { }
+
+      /**
+       * Add one element of type S to array, promoting to type T.
+       *
+       * @param u element to add
+       * @returns this array_builder object
+       */
+      template <typename S>
+      array_builder& add(const S& u) {
+        x_.push_back(promote_elements<T, S>::promote(u));
+        return *this;
+      }
+
+      /**
+       * Getter method to return array itself.
+       *
+       * @returns std:vector<T> of composed array elements.
+       */
+      std::vector<T> array() {
```

Comment:
We're talking about `array_builder` method `array()`.  The question is whether the `array_builder` instance 

* holds a reference to the returned object as a member variable;  if it's only local, you can't return a reference, and

* persists through the lifespan of the returned reference

You can always copy on return.  For instance, if you have a function

```
const foo& bar();
```

then you can copy on use with:

```
foo x = bar();
```

or you can just preserve the reference without copying as

```
const foo& x = bar();
```

In the latter case, the reference source must persist through the use of `x`.
And of course, you can always copy on return from the outside.  For example, if I have